### ISSUE TITLE
for flat classpath should we try put dependencies in a separate directory by default?

### ISSUE THREAD

**Issue by jstrachan** [fabric8io/docker-maven-plugin#535]
I saw this blog recently which was discussing why fat jars can be suboptimal:
http://product.hubspot.com/blog/the-fault-in-our-jars-why-we-stopped-building-fat-jars

each build requires re-downloading all jars again; even when no dependencies change.

I wonder if there's a way to promote base image sharing with flat classpath apps? e.g. if folks are doing CI/CD and doing lots of builds per day of the same microservice; most times its only the microservices's code changing; the dependencies will rarely change. So it'd be awesome if one docker layer was the dependencies; then another docker layer was the apps jar. Then this would promote docker caching.

I wonder if we could implement this by default by using a dependencies folder? e.g. for microservice foo we create an install like this...

```
/app
  foo.jar
  dependencies/
     jackson-core-2.7.4.jar
     ...
```

then the Dockerfile contains 2 ADD statements; one for the dependencies, one for the builds jar. 

Then hopefully docker will try cache the dependency layer as much as possible; which will help speed up docker builds, pushes, pulls

**Comment by rhuss** [comment]
When we use flat classpath files (which really makes sense for speeding up deployments), then plainf caching should be already easily possible, if not already working. One would need to ensure the proper order of `ADD` commands used, so that dependencies come first. 

The alternative would be to operate with two images: One base image holding the dependencies and a sub-image holding only the application code. This would work even without relying on Docker caching to work. This approach could probably already done (since d-m-p supports already multiple images), although they will be built again for each run (but then also probably already fully cached when nothing changes). 

But working with one image is probably easier, so let's have a look how we add things today.

Also, in order to fully replace the previous hawt-app functionality something like #257 would be nice (although it could be realized with the dependency plugin, too, it makes sense add a simple classpath file generation from a classpath file, too).

**Comment by jstrachan** [comment]
I watched this demo today; its a kinda argument for app servers (as in you only rebuild / deploy your WAR and they tend to be really tiny as if you stick to JEE then there's no dependencies in your WAR)
https://www.youtube.com/watch?v=aZH01HzUlww

its certainly a compelling argument; though personally its not quite compelling enough given the flexibility, simplicity & awesomeness of spring boot :)

So adding files individually sounds like it could make gazillions of images if a typical app has around 40 transitive dependencies and you have many images? Though if they are sorted in a good order; folks could share some base images across apps I guess (at the expense of docker image explosion).

I'm wondering if having a single image for dependencies; then separate image for the app may be a little easier to understand & reduce load on the docker registries (and pushing/pulling time).

Making 2 docker images is not so hard; making image `foo-dependencies`, then `foo` which is `FROM foo-dependencies` etc. The tricky bit is figuring out a nice way to try reuse the built`foo-dependencies` image by default so the build is nice and fast and to only rebuild it if the dependencies really have changed. 

Maybe we create a single docker label value for all the mvn coords as a single sorted string; then query all the `foo-dependencies` images for that label value; if so use it - otherwise make a new image? Wonder how long a docker label can be; the GAV coordinates could be the docker image label? :)

**Comment by rhuss** [comment]
Yeah, I've seen Adam live propagating his 'slim' Java EE approach. And well, he is indeed somehow right, that the stuff which gets moved around all the time is the tiny war file, whereas the fat dependencies are baked into the app-server base image.

For spring-boot, I don't know whether they support a flat classpath model (have only seen the fat jar thing). For fat-jar there is no easy way to optimise.

For flat classpath we indeed can go for two images for plain docker builds, not for s2i I'm afraid. For geneators we already could create two image configurations which should be build, but we would need to distinguish at that point already whether we are building with Docker or with S2I (because for the later only a single docker.tar works). 

In order to decide whether the dependencies image should be build requires now also extra logic in d-m-p (if we don't simply rely on Docker layer caching). Would love to avoid this extra step, so maybe we could try this first without.

w.r.t. to tagging alls deps as labels: whats about using a simple checksum over all coords as label ?

**Comment by jstrachan** [comment]
Agreed - fat jars should die. We should be able to just use a flat classloader using a dependencies folder + the apps jar really.

Ah good point on s2i - we'd need 2 S2I builds too ;) ouch.

I guess we could experiment with docker caching; it might be enough. But it'd be nice to avoid even making a tar of 40 jars to POST to docker just for it to tell us there's no build required (which is gonna happen for 98% of all builds).

Checksum sounds good too

**Comment by rhuss** [comment]
i see, Docker cache is only good for avoiding lengthy `RUN` commands to be called, not so for `ADD`ing stuff and as you said one would need to POST the 50MB+ anyway. So a client side check would safe quite a bit, agreed.

**Comment by jstrachan** [comment]
how about if by default we always `ADD` the dependencies first in the generated Dockerfile; then we `ADD` the 'skinny jar / war' second? (If it makes things easier lets use 2 different folders for dependencies and the 'skinny jar/war')

Then when we build in `interactive developer mode` we calculate a hash/string of all the dependencies and versions; build the image as normal - then we figure out the docker layer (docker image ID) created by adding the dependencies. (This is the trickiest bit really).

Then when we do a second build, we see the hash/string and docker layer - see that dependencies haven't changed; so the Dockerfile we generate ignores everything from the 'ADD dependencies' upwards and starts `FROM` the previous dependencies image and just does the `ADD` of the skinny jar/war. So we effectively split the generated Dockerfile logically into 2 parts; and only rebuild the first part if the dependencies change.

That way we don't have to confuse the developer into creating 2 logical docker images for each build; we just reuse the docker layers.  But it lets us optimise away having to jar up 50Mb of dependencies and POST them to docker just for docker to tell us its cacheable - when we know its cached as we previously built it ;)

**Comment by chirino** [comment]
Keep in mind the dev optimization could just build from the previous build Image if non of the deps have changed.

So in a normal build we only use the tradition 1 Docker file, and only in the RAD case do we use a 2nd Docker file that builds against the previous image.  Still do have to track dependency change on the client side.

**Comment by jstrachan** [comment]
actually thats true - we could just use final first docker image as the base for all future builds (provided we don't change the name of the tiny jar/war) - then we don't need any complex logic finding the right layer to start from etc

**Comment by chirino** [comment]
this also has a chance of working with s2i binary deployments.

**Comment by jstrachan** [comment]
@chirino yeah it should totally work with docker or S2I binary builds; since its just an option to generate a smaller  incremental Dockerfile that just uses the first clean build's image output and then just copy the current project's artefact
